### PR TITLE
Fix followers checkpoint race

### PR DESCRIPTION
**Pull request by eolivelli**
- Fix race between follower thread and checkpoint thread
- Graceful exit of follower procedure during server shutdown
- Use better API to read from BookKeeper (less memory copies, less allocations, reading directly from the Netty Buffer)
- Clean up logs
- Add system properties for fine tuning
- By default recovery will read now 100.000 entries and not 10.000 from BK
- By default the follower will read batches of 10.000 entries from BK
- Add a simple integration test that forces checkpoints of a follower

### CODE REVIEW

File: herddb-core/src/main/java/herddb/cluster/BookkeeperCommitLog.java [hunk 1]
```diff
@@ -432,54 +440,62 @@ public void recovery(
...
-                                currentLedgerId = ledgerId;
-                                lastSequenceNumber.set(entryId);
-                                if (number.after(snapshotSequenceNumber)) {
-                                    LOGGER.log(Level.FINEST, "RECOVER ENTRY #" + localEntryCount + " {0}, {1}",
-                                            new Object[]{number, statusEdit});
-                                    consumer.accept(number, statusEdit);
-                                } else {
-                                    LOGGER.log(Level.FINEST, "SKIP ENTRY #" + localEntryCount + " {0}<{1}, {2}",
-                                            new Object[]{number, snapshotSequenceNumber, statusEdit});
+                            try (LedgerEntries entries = handle.read(start, end);) {
```
Comment (eolivelli) [inline herddb-core/src/main/java/herddb/cluster/BookkeeperCommitLog.java]: using BK 4.8+ API, with direct access to Netty ByteBuf

File: herddb-core/src/main/java/herddb/cluster/BookkeeperCommitLog.java [hunk 2]
```diff
@@ -770,7 +786,11 @@ public void followTheLeader(
                     readLastAddConfirmedAndEntry(nextEntry, LONG_POLL_TIMEOUT, false)) {
                 if (entryAndLac.hasEntry()) {
                     org.apache.bookkeeper.client.api.LedgerEntry e = entryAndLac.getEntry();
-                    acceptEntryForFollower(e, consumer);
+                    boolean canContinue = acceptEntryForFollower(e, consumer);
```
Comment (eolivelli) [inline herddb-core/src/main/java/herddb/cluster/BookkeeperCommitLog.java]: graceful exit

File: herddb-core/src/main/java/herddb/cluster/BookkeeperCommitLog.java [hunk 3]
```diff
@@ -793,26 +813,29 @@ public void followTheLeader(
...
-    private void acceptEntryForFollower(
+    private boolean acceptEntryForFollower(
             org.apache.bookkeeper.client.api.LedgerEntry e,
-            BiConsumer<LogSequenceNumber, LogEntry> consumer
-    ) throws EOFException {
+            EntryAcceptor consumer
+    ) throws Exception {
         long entryId = e.getEntryId();
-        byte[] entryData = e.getEntryBytes();
-        LogEntry statusEdit = LogEntry.deserialize(entryData);
```
Comment (eolivelli) [inline herddb-core/src/main/java/herddb/cluster/BookkeeperCommitLog.java]: save some allocations and read directly from Netty ByteBuf

File: herddb-core/src/main/java/herddb/core/TableSpaceManager.java [hunk 4]
```diff
@@ -976,13 +976,19 @@ public String toString() {
...
                 while (!isLeader() && !closed) {
+                    long readLock = acquireReadLock("follow");
+                    try {
                     log.followTheLeader(actualLogSequenceNumber, (LogSequenceNumber num, LogEntry u) -> {
                         try {
                             apply(new CommitLogResult(num, false, true), u, false);
                         } catch (Throwable t) {
                             throw new RuntimeException(t);
                         }
+                        return !isLeader() && !closed;
```
Comment (eolivelli) [inline herddb-core/src/main/java/herddb/core/TableSpaceManager.java]: gracefull exit, related to #536

File: herddb-core/src/main/java/herddb/core/TableSpaceManager.java [hunk 5]
```diff
@@ -976,13 +976,19 @@ public String toString() {
         public void run() {
             try (CommitLog.FollowerContext context = log.startFollowing(actualLogSequenceNumber)) {
                 while (!isLeader() && !closed) {
+                    long readLock = acquireReadLock("follow");
```
Comment (diegosalvi) [inline herddb-core/src/main/java/herddb/core/TableSpaceManager.java]: This is the actual fix

**Comment by aluccaroni** [comment]
@eolivelli  @diegosalvi  this pull will make herddb incompatibile with BK < 4.8?

**Comment by eolivelli** [comment]
@aluccaroni no
It is still compatible with bookies up to 4.5 :)

**Comment by eolivelli** [comment]
@aluccaroni we were already using the BK 4.8 API for the "follower" part (LedgerEntries) but not for the RecoveryPart.

btw it is only a client side API, no changes to the wire protocol, it is only a cleaner Java API, that enables the application to control when memory is released 
```
try (LedgerEntries ...) {
}  // here Java adds a LedgerEntries.close() that releases Netty/Direct memory explicly

```

**Comment by eolivelli** [comment]
With this fix we can close #536

**Comment by diegosalvi** [comment]
Spotbug fails

**Pull request merged**